File: zone-server/src/main/kotlin/net/bestia/zone/world/ground/GroundLayer.kt
package net.bestia.zone.world.ground

/**
 * A kind of lasting mark on the ground, and the channel the client composites it through.
 *
 * ### Why the ground needs layers at all
 *
 * Burnt, worn, bloodied and trodden are the same question asked about the same lattice, so they share one
 * store, one message and one decay pass. What they do not share is where their mask comes from: [SCORCHED] is
 * written by the fire, [WORN] accumulates under traffic, [BLOODIED] is stamped once where something died.
 * That difference lives in the writers, not here.
 *
 * ### None of this can be a `BlockType`
 *
 * `CHUNK_PATCH_ENCODING_REMOVAL_V1` is the only patch encoding the protocol has, so no message can change a
 * voxel's material - a client derives a patched voxel's material locally. Ground history therefore travels
 * beside the ground rather than in it, which is the same argument `ChunkGroundOverlaySMSG` already makes for
 * scorch.
 *
 * ### [channel] is a wire contract
 *
 * It is both the RGBA channel the client packs this layer into and the order the layers composite in, lowest
 * first. Reordering them silently repaints the world, so this is append-only and pinned by `GroundLayerTest`
 * on this side and a fixture on the client's. [wireId] is separate and never reused, so a layer can be
 * retired without renumbering the ones after it.
 */
enum class GroundLayer(val wireId: Int, val channel: Int) {

  /** Ground a fire has been through. Written by the fire, healed by rain. */
  SCORCHED(wireId = 1, channel = 0),

  /** Ground walked bare. The only layer that accumulates rather than being set. */
  WORN(wireId = 2, channel = 1),

  /** Spilled where something died, and the reason a battle site still reads as one a day later. */
  BLOODIED(wireId = 3, channel = 2),

  /**
   * Ground broken open by something passing: prints in snow, in loose sand, in mud.
   *
   * Composites last and differently from the three above - it does not paint a material, it *cuts into* the
   * weather snow the shader already derives. So it is only visible where there is something to disturb,
   * which is why a dusting of snow makes tracks legible that the same ground shows nothing of in summer.
   */
  DISTURBED(wireId = 4, channel = 3);

  companion object {

    /** How many channels one mark texture carries, which is what caps the layers that may stack on a cell. */
    const val CHANNELS = 4

    fun ofWireId(wireId: Int): GroundLayer? {
      return entries.firstOrNull { it.wireId == wireId }
    }
  }
}

File: bestia-client/src/Game/World/GroundLayer.cs
namespace BestiaBehemothClient.Game.World
{
  /// <summary>
  /// A kind of lasting mark on the ground, and the channel it composites through.
  /// </summary>
  /// <remarks>
  /// The mirror of the server's <c>GroundLayer</c>. Both numbers are wire contracts and neither is checkable
  /// at runtime - any byte string is a legal payload, so a disagreement paints the world wrong rather than
  /// throwing. <c>GroundLayerTest</c> pins them on both sides.
  ///
  /// <para>
  /// A mark cannot be a <c>BlockType</c>: the chunk patch format can only ever remove a voxel, so no message
  /// in this protocol can change a voxel's material. Ground history travels beside the ground.
  /// </para>
  /// </remarks>
  public static class GroundLayers
  {
    /// <summary>
    /// How many marks may stack on one square metre, which is the channels one mark texture carries.
    /// </summary>
    /// <remarks>
    /// Four, because the mark texture is a single <c>RGBA8</c>. Unlike
    /// <see cref="Mesh.BlockAppearance.Slots"/> this is not near an engine limit - a second texture buys four
    /// more for one more fetch - so it is a budget rather than a wall.
    /// </remarks>
    public const int Channels = 4;

    /// <summary>Wire ids, as the proto's <c>GroundLayerId</c> numbers them. Never reused.</summary>
    public enum Id
    {
      Unspecified = 0,
      Scorched = 1,
      Worn = 2,
      Bloodied = 3,
      Disturbed = 4
    }

    /// <summary>
    /// Which channel a layer packs into, and the order layers composite in, lowest first.
    /// </summary>
    /// <returns>-1 for a layer this build does not know, which is skipped rather than guessed at.</returns>
    public static int ChannelOf(Id layer) => layer switch
    {
      Id.Scorched => 0,
      Id.Worn => 1,
      Id.Bloodied => 2,
      Id.Disturbed => 3,
      _ => -1
    };
  }
}
